megablocks/ops/padded_scatter.py

```diff
@@ -20,18 +32,36 @@ def forward(ctx, x, indices, bin_ids, weights, bins, padded_bins, top_k):
     def backward(ctx, grad):
         grad = grad.contiguous()
 
-        x, indices, bin_ids, weights, bins, padded_bins = ctx.saved_tensors
-        out = kernels.padded_gather(
-            grad,
-            indices,
-            bin_ids,
-            weights,
-            bins,
-            padded_bins,
-            ctx.top_k)
+        indices, bin_ids, weights, bins, padded_bins = ctx.saved_tensors[-5:]
+        dgrad = None
+        if ctx.needs_input_grad[0]:
+            dgrad = kernels.padded_gather(
+                grad,
+                indices,
+                bin_ids,
+                weights,
+                bins,
+                padded_bins,
+                ctx.top_k)
 
         wgrad = None
-        if ctx.needs_input_grad[3]:
+        if ctx.needs_input_grad[3]:  # need wgrad
+            if ctx.num_bits == -1:
+                x = ctx.saved_tensors[0]
+            else:
+                if len(ctx.saved_tensors) != 7:
```

Comment:
What is going on here?

Reply:
I got a weird error indicating that only one tensor got saved in the forward pass when the backward pass was expecting both the quantized input and the scales. I _think_ I fixed this by just checking for `ctx.needs_input_grad[3]` in `saved_x = x if ctx.needs_input_grad[3] else None` instead of for this and `weight is not None`. But I could never repro the error so I added this to at least throw something more informative if it ever happens again. A little ugly and not certain it's needed, so happy to take this block out if you don't like it.

Reply:
Interesting. I'd say lets remove this if you've never repro'd it and we can add it back if this issue returns.